core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -188,6 +191,17 @@ public ZenDiscovery(Settings settings, ThreadPool threadPool, TransportService t
             DISCOVERY_REJOIN_ACTION_NAME, RejoinClusterRequest::new, ThreadPool.Names.SAME, new RejoinClusterRequestHandler());
     }
 
+    // protected to allow overriding in tests
+    protected ZenPing newZenPing(Settings settings, ThreadPool threadPool, TransportService transportService,
+                                 UnicastHostsProvider hostsProvider) {
+        return new UnicastZenPing(settings, threadPool, transportService, hostsProvider);
+    }
+
+    // TODO: this shoudln't be public, clean up the zen ping interface somehow to better allow clearing for testing?
```

Comment:
can't we be build this into the mock discovery and have a setting on it on whether it should use mock pings or not?